types/validation.go

```diff
@@ -1,30 +1,141 @@
-func ValidateTime(t time.Time) error {
-	var (
-		now     = tmtime.Now()
-		oneYear = 8766 * time.Hour
-	)
-	if t.Before(now.Add(-oneYear)) || t.After(now.Add(oneYear)) {
-		return fmt.Errorf("time drifted too much. Expected: -1 < %v < 1 year", now)
+// VerifyCommit verifies +2/3 of the set had signed the given commit.
+//
+// It checks all the signatures! While it's safe to exit as soon as we have
+// 2/3+ signatures, doing so would impact incentivization logic in the ABCI
+// application that depends on the LastCommitInfo sent in BeginBlock, which
+// includes which validators signed. For instance, Gaia incentivizes proposers
+// with a bonus for including more than +2/3 of the signatures.
+func VerifyCommit(chainID string, vals *ValidatorSet, blockID BlockID,
+	height int64, commit *Commit) error {
+
+	// run a basic validation of the arguments
+	if err := verifyBasicValsAndCommit(vals, commit, height, blockID); err != nil {
+		return err
 	}
-	return nil
+
+	// calculate voting power needed. Note that total voting power is capped to
+	// 1/8th of max int64 so this operation should never overflow
+	votingPowerNeeded := vals.TotalVotingPower() * 2 / 3
+
+	// ignore all absent signatures
+	ignore := func(c CommitSig) bool { return c.Absent() }
+
+	// only count the signatures that are for the block
+	count := func(c CommitSig) bool { return c.ForBlock() }
+
+	// attempt to batch verify
+	cacheSignBytes, success, err := tryVerifyCommitBatch(
+		chainID, vals, commit, votingPowerNeeded, ignore, count, true, true)
+	if err != nil {
+		return err
+	}
+	if success {
+		return nil
+	}
+
+	// if verification failed or is not supported then fallback to single verification
+	return verifyCommitSingle(chainID, vals, commit, votingPowerNeeded,
+		cacheSignBytes, ignore, count, true, true)
+}
+
+// LIGHT CLIENT VERIFICATION METHODS
+
+// VerifyCommitLight verifies +2/3 of the set had signed the given commit.
+//
+// This method is primarily used by the light client and does not check all the
+// signatures.
+func VerifyCommitLight(chainID string, vals *ValidatorSet, blockID BlockID,
+	height int64, commit *Commit) error {
+
+	// run a basic validation of the arguments
+	if err := verifyBasicValsAndCommit(vals, commit, height, blockID); err != nil {
+		return err
+	}
+
+	// calculate voting power needed
+	votingPowerNeeded := vals.TotalVotingPower() * 2 / 3
+
+	// ignore all commit signatures that are not for the block
+	ignore := func(c CommitSig) bool { return !c.ForBlock() }
+
+	// count all the remaining signatures
+	count := func(c CommitSig) bool { return true }
+
+	// attempt to batch verify
+	cacheSignBytes, success, err := tryVerifyCommitBatch(
+		chainID, vals, commit, votingPowerNeeded, ignore, count, false, true)
+	if err != nil {
+		return err
+	}
+	if success {
+		return nil
+	}
+
+	// if verification failed or is not supported then fallback to single verification
+	return verifyCommitSingle(chainID, vals, commit, votingPowerNeeded,
```

Comment:
@marbar3778 Do you think for verify light that we need to fall back to single verification if batch verification fails. I mean if one of the signatures are invalid, then wouldn't we consider the entire commit to be invalid. Maybe it's a minor point

Reply:
This is true. I don't see this happening often, so we should be fine with going with this path. If we begin observing this happening then we can reconsider how it is done. 

if 1/3+ fails we would consider it invalid, since we only check that 2/3+ is valid here.